docs/manual/src/internals/design_principles.md

```diff
@@ -0,0 +1,95 @@
+# Design Principles
+
+These are some high-level points to consider when making changes to UniFFI (or when wondering why past changes were made in a particular way).
+
+
+### Prioritize Mozilla's short-term needs
+
+The initial consumers of this tool are teams working on features for Mozilla's mobile browsers.
+While we try to make the tool generally useful, we'll invest first in things that are the most valuable
+to those teams, which are reflected in the points below.
+
+
+### Safety First
+
+The generated bindings need to be safe by default. It should be impossible for foreign-language code
+to trigger undefined behaviour in Rust by calling the public API of the generated bindings, even if it
+is called in egregiously wrong or malicious ways. We will accept reduced performance in the interests
+of ensuring this safety.
+
+(The meaning of "impossible" and "public API" will of course depend on the target language. For example,
+code in Python might mutate internal attributes of an object that are marked as private with a leading
+underscore, and there's not much we can do to guard against that.)
+
+Where possible, we use Rust's typesystem to encode safety guarantees. If that's not possible then the
+generated Rust code may use `unsafe` and assume that the generated foreign-language code will uphold
+safety guarantees at runtime.
+
+**Example:** We insist that all object instances exposed to foreign-language code be `Sync` and `Send`,
+so that they're safe to access regardless of the threading model of the calling code. We do not allow
+thread-safety guarantees to be deferred to assumptions about how the code is called.
+
+**Example:** We do not allow returning any borrowed data from function calls, because we can't make
+any guarantees about when or how the foreign-language could would access it.
+
+
+### Performance is a feature, but not a deal-breaker
+
+Our initial use-cases are not performance-critical, and our team are not low-level Rust experts,
+so we're highly motivated to favour simplicity and maintainability over performance. Given the choice
+we will pick "simple but slow" over "fast but complicated".
+
+However, we know that performance can degrade through thousands of tiny cuts, so we'll keep iterating
+towards the winning combination of "simple *and* fast" over time.
+
+**Example:** Initial versions of the tool used opaque integer handles and explicit mutexes to manage
+object references, favouring simplicity (in the "we're confident this works as intended" sense) over
+performance. As we got more experience and confidence with the approach tool we replaced handles with
+raw `Arc` pointers, which both simplified the code and removed some runtime overheads.
+
+**Violation:** Initial versions of the tool pass structured data over the FFI by serializing it to
```

Comment:
The use on `Initial versions` here makes it sound like "in the past", where this is still true today - whereas in the example above, it truly is now in the past. I think it's probably OK to say something like "at time of writing ..." to clear that up.